invenio_records_resources/services/records/queryparser/query.py

```diff
@@ -82,13 +114,23 @@ def parse(self, query_str):
             # We parse the Lucene query syntax in Python, so we know upfront
             # if the syntax is correct before executing it in the search engine
             tree = luqum_parser.parse(query_str)
-
             # Perform transformation on the abstract syntax tree (AST)
-            if self.tree_transformer_factory is not None:
-                transformer = self.tree_transformer_factory()
+            if self.tree_transformer_cls is not None:
+                transformer = self.tree_transformer_cls(
+                    mapping=self.mapping, allow_list=self.allow_list
+                )
                 new_tree = transformer.visit(tree, context={"identity": self.identity})
                 query_str = str(new_tree)
             return dsl.Q("query_string", query=query_str, **self.extra_params)
         except (ParseError, QuerystringValidationError):
             # Fallback to a multi-match query.
+            if self.allow_list:
+                # if there is an allow list it must overwrite a potential value
+                # given by the query to include it in the fields
+                kwargs = {**self.extra_params, "fields": self.fields}
```

Comment:
that's different from `self._fields`?

Reply:
Yes, `self.fields` is a property that is calculated using the original `_fields`, adding the allowed list, and removing duplicates. Then depending on the case (i.e. the if in line 127) we might need to use the calculated fields or simply the raw original ones. That's why we do not modify them (they are still in extra_params).

The reason to use `_fields` and `fields` is to avoid doing the calculation in the constructor, since it is not always needed. See first review comments form manuel.

On top of that we extract from extra_params to a private attribute because we do not want to modify the values stored in extra params

Reply:
It's dummy but could we rename it to `calculated_fields`? or have a detailed docstring on what are the differences between the `self.fields` and `self._fields`?

Reply:
I've added an explanation on the docstring and there is an explanation of the actual calculations in line 102